File: android/app/src/main/java/jp/yojiro/sakelog/SakeVisionPlugin.kt
package jp.yojiro.sakelog

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.BatteryManager
import android.os.Build
import android.os.PowerManager
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.io.File
import java.util.Collections
import java.util.concurrent.Executors

@CapacitorPlugin(name = "SakeVision")
class SakeVisionPlugin : Plugin() {
    private val executor = Executors.newSingleThreadExecutor()
    private val cancelled = Collections.synchronizedSet(mutableSetOf<String>())

    @PluginMethod
    fun getCapabilities(call: PluginCall) {
        val power = context.getSystemService(PowerManager::class.java)
        val battery = context.getSystemService(BatteryManager::class.java)
        call.resolve(JSObject().apply {
            put("environment", "android-native")
            put("platform", "Android")
            put("osVersion", Build.VERSION.RELEASE)
            put("deviceModel", "${Build.MANUFACTURER} ${Build.MODEL}")
            put("ocrEngine", "mlkit")
            put("barcodeEngine", "ML Kit Barcode Scanning 17.3.0")
            put("visualEngine", "sake-local-fingerprint-v1")
            put("modelVersion", "ML Kit Text Recognition 16.0.1 bundled")
            put("supportsLabelDetection", true)
            put("supportsTextRecognition", true)
            put("supportsBarcode", true)
            put("supportsVisualEmbedding", true)
            put("thermalState", if (Build.VERSION.SDK_INT >= 29) power.currentThermalStatus.toString() else "unavailable")
            put("batteryLevel", battery.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY) / 100.0)
        })
    }

    @PluginMethod
    fun cancel(call: PluginCall) {
        call.getString("signalId")?.let { cancelled.add(it) }
        call.resolve()
    }

    @PluginMethod
    fun detectLabelRegions(call: PluginCall) = withBitmap(call) { bitmap, started ->
        JSObject().apply {
            put("regions", regionArray(bitmap))
            put("processingTimeMs", elapsed(started))
        }
    }

    @PluginMethod
    fun recognizeText(call: PluginCall) = withBitmap(call) { bitmap, started ->
        val observations = recognize(bitmap, call.getString("signalId"))
        JSObject().apply {
            put("observations", observations)
            put("processingTimeMs", elapsed(started))
            put("engine", "mlkit")
        }
    }

    @PluginMethod
    fun readBarcodes(call: PluginCall) = withBitmap(call) { bitmap, started ->
        JSObject().apply {
            put("observations", barcodes(bitmap))
            put("processingTimeMs", elapsed(started))
        }
    }

    @PluginMethod
    fun createImageEmbedding(call: PluginCall) = withBitmap(call) { bitmap, started ->
        val values = NativeVisionModels.embedding(bitmap)
        JSObject().apply {
            put("values", NativeVisionModels.toArray(values))
            put("model", "sake-local-fingerprint-v1")
            put("dimensions", values.size)
            put("processingTimeMs", elapsed(started))
        }
    }

    @PluginMethod
    fun compareImages(call: PluginCall) {
        executor.execute {
            try {
                val left = loadBitmap(call.getString("firstFileUri") ?: error("firstFileUri is required"))
                val right = loadBitmap(call.getString("secondFileUri") ?: error("secondFileUri is required"))
                val result = NativeVisionModels.similarity(NativeVisionModels.embedding(left), NativeVisionModels.embedding(right))
                left.recycle(); right.recycle()
                call.resolve(JSObject().apply { put("distance", result.first); put("similarity", result.second); put("model", "sake-local-fingerprint-v1") })
            } catch (error: Exception) { call.reject("画像特徴を比較できませんでした。", error) }
        }
    }

    @PluginMethod
    fun analyzeImage(call: PluginCall) = withBitmap(call) { bitmap, started ->
        val signalId = call.getString("signalId")
        if (signalId != null && cancelled.contains(signalId)) error("CANCELLED")
        val labels = regionArray(bitmap)
        val text = recognize(bitmap, signalId)
        if (signalId != null && cancelled.contains(signalId)) error("CANCELLED")
        JSObject().apply {
            put("textObservations", text)
            put("barcodeObservations", barcodes(bitmap))
            put("labelRegions", labels)
            put("visualEmbedding", JSObject().apply {
                val values = NativeVisionModels.embedding(bitmap)
                put("values", NativeVisionModels.toArray(values)); put("model", "sake-local-fingerprint-v1"); put("dimensions", values.size)
            })
            put("imageQuality", JSObject().apply {
                put("blurScore", 0.5); put("brightnessScore", 0.5); put("glareScore", 0.0)
                put("labelCoverage", if (labels.length() > 0) 0.4 else 0.0); put("warnings", JSArray())
            })
            put("processingTimeMs", elapsed(started))
            put("warnings", JSArray())
        }
    }

    private fun recognize(bitmap: Bitmap, signalId: String?): JSArray {
        val regions = NativeVisionModels.detectLabelRegions(bitmap)
        val passes = regions.mapIndexed { index, rect -> Bitmap.createBitmap(bitmap, rect.left, rect.top, rect.width(), rect.height()) to "label-$index" } + listOf(bitmap to "full")
        val output = JSArray()
        val recognizers = listOf(
            "ja" to TextRecognition.getClient(JapaneseTextRecognizerOptions.Builder().build()),
            "latin" to TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        )
        try {
            passes.forEachIndexed { passIndex, (imageBitmap, passId) ->
                if (signalId != null && cancelled.contains(signalId)) error("CANCELLED")
                recognizers.forEach { (language, recognizer) ->
                    val result = Tasks.await(recognizer.process(InputImage.fromBitmap(imageBitmap, 0)))
                    appendText(output, result, bitmap.width, bitmap.height, if (passIndex < regions.size) "frontLabel" else "fullImage", "$passId-$language")
                }
                if (imageBitmap !== bitmap) imageBitmap.recycle()
            }
        } finally { recognizers.forEach { it.second.close() } }
        return output
    }

    private fun appendText(output: JSArray, text: Text, width: Int, height: Int, regionType: String, passId: String) {
        text.textBlocks.forEach { block -> block.lines.forEach { line -> line.elements.forEach { element ->
            val box = element.boundingBox ?: return@forEach
            output.put(JSObject().apply {
                put("text", element.text); put("confidence", element.confidence.toDouble())
                put("boundingBox", NativeVisionModels.normalizedRect(box, width, height))
                put("cornerPoints", JSArray()); put("language", element.recognizedLanguage)
                put("orientation", 0); put("regionType", regionType); put("engine", "mlkit"); put("passId", passId)
            })
        } } }
    }

    private fun barcodes(bitmap: Bitmap): JSArray {
        val options = BarcodeScannerOptions.Builder().setBarcodeFormats(
            Barcode.FORMAT_EAN_8, Barcode.FORMAT_EAN_13, Barcode.FORMAT_UPC_A, Barcode.FORMAT_UPC_E,
            Barcode.FORMAT_CODE_128, Barcode.FORMAT_QR_CODE, Barcode.FORMAT_DATA_MATRIX
        ).build()
        val scanner = BarcodeScanning.getClient(options)
        return try {
            val values = Tasks.await(scanner.process(InputImage.fromBitmap(bitmap, 0)))
            JSArray().apply { values.forEach { barcode -> barcode.rawValue?.let { value -> put(JSObject().apply {
                put("rawValue", value); put("format", barcode.format.toString()); put("confidence", 1.0)
                barcode.boundingBox?.let { put("boundingBox", NativeVisionModels.normalizedRect(it, bitmap.width, bitmap.height)) }
            }) } } }
        } finally { scanner.close() }
    }

    private fun regionArray(bitmap: Bitmap): JSArray = JSArray().apply {
        NativeVisionModels.detectLabelRegions(bitmap).forEachIndexed { index, rect -> put(JSObject().apply {
            put("id", "android-region-$index"); put("boundingBox", NativeVisionModels.normalizedRect(rect, bitmap.width, bitmap.height))
            put("cornerPoints", JSArray()); put("confidence", 0.55); put("regionType", if (index == 0) "frontLabel" else "backLabel"); put("perspectiveCorrected", false)
        }) }
    }

    private fun withBitmap(call: PluginCall, operation: (Bitmap, Long) -> JSObject) {
        val uri = call.getString("localFileUri") ?: run { call.reject("localFileUri is required"); return }
        executor.execute {
            var bitmap: Bitmap? = null
            try {
                val started = System.nanoTime()
                bitmap = loadBitmap(uri)
                call.resolve(operation(bitmap, started))
            } catch (error: Exception) { call.reject(if (error.message == "CANCELLED") "画像解析をキャンセルしました。" else "画像解析に失敗しました。", error) }
            finally { bitmap?.recycle() }
        }
    }

    private fun loadBitmap(value: String): Bitmap {
        val uri = Uri.parse(value)
        return if (uri.scheme == "content") {
            context.contentResolver.openInputStream(uri).use { BitmapFactory.decodeStream(it) ?: error("画像をデコードできませんでした") }
        } else {
            val path = if (uri.scheme == "file") uri.path else value
            BitmapFactory.decodeFile(File(path ?: value).absolutePath) ?: error("画像をデコードできませんでした")
        }
    }

    private fun elapsed(started: Long) = (System.nanoTime() - started) / 1_000_000.0

    override fun handleOnDestroy() {
        executor.shutdownNow()
        super.handleOnDestroy()
    }
}

File: ios/App/App/SakeVisionPlugin.swift
import Foundation
import UIKit
import Vision
import CoreImage
import Capacitor

@objc(SakeVisionPlugin)
public class SakeVisionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SakeVisionPlugin"
    public let jsName = "SakeVision"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "analyzeImage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectLabelRegions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "recognizeText", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readBarcodes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createImageEmbedding", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "compareImages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise)
    ]

    private let queue = DispatchQueue(label: "jp.yojiro.sakelog.vision", qos: .userInitiated)
    private var cancelled = Set<String>()
    private let cancelLock = NSLock()

    @objc func getCapabilities(_ call: CAPPluginCall) {
        UIDevice.current.isBatteryMonitoringEnabled = true
        var capabilities: [String: Any] = [
            "environment": "ios-native",
            "platform": "iOS",
            "osVersion": UIDevice.current.systemVersion,
            "deviceModel": UIDevice.current.model,
            "ocrEngine": "apple-vision",
            "barcodeEngine": "VNDetectBarcodesRequest",
            "visualEngine": "VNGenerateImageFeaturePrintRequest",
            "modelVersion": "Apple Vision / iOS \(UIDevice.current.systemVersion)",
            "supportsLabelDetection": true,
            "supportsTextRecognition": true,
            "supportsBarcode": true,
            "supportsVisualEmbedding": true,
            "thermalState": thermalStateName(ProcessInfo.processInfo.thermalState)
        ]
        if UIDevice.current.batteryLevel >= 0 {
            capabilities["batteryLevel"] = Double(UIDevice.current.batteryLevel)
        }
        call.resolve(capabilities)
    }

    @objc func cancel(_ call: CAPPluginCall) {
        if let signalId = call.getString("signalId") {
            cancelLock.lock(); cancelled.insert(signalId); cancelLock.unlock()
        }
        call.resolve()
    }

    @objc func detectLabelRegions(_ call: CAPPluginCall) {
        withImage(call) { image, started in
            let regions = try self.detectRegions(image)
            return ["regions": regions, "processingTimeMs": self.elapsed(started)]
        }
    }

    @objc func recognizeText(_ call: CAPPluginCall) {
        withImage(call) { image, started in
            let signalId = call.getString("signalId")
            let regions = try self.detectRegionObservations(image)
            let observations = try self.recognize(image, regions: regions, signalId: signalId)
            return ["observations": observations, "processingTimeMs": self.elapsed(started), "engine": "apple-vision"]
        }
    }

    @objc func readBarcodes(_ call: CAPPluginCall) {
        withImage(call) { image, started in
            let request = VNDetectBarcodesRequest()
            request.symbologies = [.ean8, .ean13, .upce, .code128, .qr, .dataMatrix]
            try VNImageRequestHandler(ciImage: image, options: [:]).perform([request])
            let values = (request.results ?? []).compactMap { item -> [String: Any]? in
                guard let payload = item.payloadStringValue else { return nil }
                return ["rawValue": payload, "format": item.symbology.rawValue, "confidence": Double(item.confidence), "boundingBox": NativeVisionModels.rect(item.boundingBox)]
            }
            return ["observations": values, "processingTimeMs": self.elapsed(started)]
        }
    }

    @objc func createImageEmbedding(_ call: CAPPluginCall) {
        withImage(call) { image, started in
            _ = try self.featurePrint(image)
            let values = self.compactEmbedding(image)
            return ["values": values, "model": "VNFeaturePrint+sake-local-fingerprint-v1", "dimensions": values.count, "processingTimeMs": self.elapsed(started)]
        }
    }

    @objc func compareImages(_ call: CAPPluginCall) {
        guard let first = call.getString("firstFileUri"), let second = call.getString("secondFileUri"),
              let firstURL = NativeVisionModels.fileURL(first), let secondURL = NativeVisionModels.fileURL(second),
              let firstImage = CIImage(contentsOf: firstURL), let secondImage = CIImage(contentsOf: secondURL) else {
            call.reject("比較する画像ファイルを読み込めませんでした。")
            return
        }
        queue.async {
            do {
                let left = try self.featurePrint(firstImage)
                let right = try self.featurePrint(secondImage)
                var distance: Float = 0
                try left.computeDistance(&distance, to: right)
                let similarity = max(0, 1 - Double(distance) / 2)
                call.resolve(["distance": Double(distance), "similarity": similarity, "model": "VNGenerateImageFeaturePrintRequest"])
            } catch { call.reject("画像特徴を比較できませんでした。", nil, error) }
        }
    }

    @objc func analyzeImage(_ call: CAPPluginCall) {
        withImage(call) { image, started in
            let signalId = call.getString("signalId")
            let regionObservations = try self.detectRegionObservations(image)
            if self.isCancelled(signalId) { throw CancellationError() }
            let text = try self.recognize(image, regions: regionObservations, signalId: signalId)
            if self.isCancelled(signalId) { throw CancellationError() }
            let barcodeRequest = VNDetectBarcodesRequest()
            barcodeRequest.symbologies = [.ean8, .ean13, .upce, .code128, .qr, .dataMatrix]
            try VNImageRequestHandler(ciImage: image, options: [:]).perform([barcodeRequest])
            let barcodes = (barcodeRequest.results ?? []).compactMap { item -> [String: Any]? in
                guard let value = item.payloadStringValue else { return nil }
                return ["rawValue": value, "format": item.symbology.rawValue, "confidence": Double(item.confidence), "boundingBox": NativeVisionModels.rect(item.boundingBox)]
            }
            let regions = regionObservations.enumerated().map { self.regionJSON($0.element, index: $0.offset) }
            let embedding = self.compactEmbedding(image)
            return [
                "textObservations": text,
                "barcodeObservations": barcodes,
                "labelRegions": regions,
                "visualEmbedding": ["values": embedding, "model": "VNFeaturePrint+sake-local-fingerprint-v1", "dimensions": embedding.count],
                "imageQuality": self.quality(image, regions: regionObservations),
                "processingTimeMs": self.elapsed(started),
                "warnings": regions.isEmpty ? ["ラベル領域を検出できなかったため画像全体も解析しました。"] : []
            ]
        }
    }

    private func withImage(_ call: CAPPluginCall, operation: @escaping (CIImage, CFAbsoluteTime) throws -> [String: Any]) {
        guard let value = call.getString("localFileUri"), let url = NativeVisionModels.fileURL(value), let image = CIImage(contentsOf: url, options: [.applyOrientationProperty: true]) else {
            call.reject("画像ファイルを読み込めませんでした。")
            return
        }
        queue.async {
            let started = CFAbsoluteTimeGetCurrent()
            do { call.resolve(try operation(image, started)) }
            catch is CancellationError { call.reject("画像解析をキャンセルしました。", "CANCELLED") }
            catch { call.reject("画像解析に失敗しました。", nil, error) }
        }
    }

    private func detectRegionObservations(_ image: CIImage) throws -> [VNRectangleObservation] {
        let request = VNDetectRectanglesRequest()
        request.maximumObservations = 4
        request.minimumConfidence = 0.45
        request.minimumAspectRatio = 0.2
        request.maximumAspectRatio = 1.0
        request.minimumSize = 0.12
        request.quadratureTolerance = 30
        try VNImageRequestHandler(ciImage: image, options: [:]).perform([request])
        return request.results ?? []
    }

    private func detectRegions(_ image: CIImage) throws -> [[String: Any]] {
        try detectRegionObservations(image).enumerated().map { regionJSON($0.element, index: $0.offset) }
    }

    private func regionJSON(_ item: VNRectangleObservation, index: Int) -> [String: Any] {
        [
            "id": "vision-region-\(index)", "boundingBox": NativeVisionModels.rect(item.boundingBox),
            "cornerPoints": NativeVisionModels.points([item.topLeft, item.topRight, item.bottomRight, item.bottomLeft]),
            "confidence": Double(item.confidence), "regionType": NativeVisionModels.regionName(index), "perspectiveCorrected": true
        ]
    }

    private func recognize(_ image: CIImage, regions: [VNRectangleObservation], signalId: String?) throws -> [[String: Any]] {
        var output: [[String: Any]] = []
        let cropPasses: [(CIImage, String)] = regions.prefix(3).enumerated().compactMap {
            guard let crop = perspectiveCrop(image, rectangle: $0.element) else { return nil }
            return (crop, "label-\($0.offset)")
        }
        let passes = cropPasses + [(image, "full")]
        for (index, pass) in passes.enumerated() {
            if isCancelled(signalId) { throw CancellationError() }
            for minimumHeight in [0.018, 0.008] {
                let request = VNRecognizeTextRequest()
                request.recognitionLevel = .accurate
                request.recognitionLanguages = ["ja-JP", "en-US"]
                request.usesLanguageCorrection = true
                request.minimumTextHeight = Float(minimumHeight)
                request.regionOfInterest = CGRect(x: 0, y: 0, width: 1, height: 1)
                try VNImageRequestHandler(ciImage: pass.0, options: [:]).perform([request])
                for observation in request.results ?? [] {
                    guard let candidate = observation.topCandidates(3).first else { continue }
                    output.append([
                        "text": candidate.string, "confidence": Double(candidate.confidence),
                        "boundingBox": NativeVisionModels.rect(observation.boundingBox),
                        "regionType": index < cropPasses.count ? NativeVisionModels.regionName(index) : "fullImage",
                        "engine": "apple-vision", "passId": "\(pass.1)-h\(minimumHeight)"
                    ])
                }
            }
        }
        return output
    }

    private func featurePrint(_ image: CIImage) throws -> VNFeaturePrintObservation {
        let request = VNGenerateImageFeaturePrintRequest()
        try VNImageRequestHandler(ciImage: image, options: [:]).perform([request])
        guard let result = request.results?.first as? VNFeaturePrintObservation else { throw NSError(domain: "SakeVision", code: 2) }
        return result
    }

    private func perspectiveCrop(_ image: CIImage, rectangle: VNRectangleObservation) -> CIImage? {
        let extent = image.extent
        func point(_ value: CGPoint) -> CIVector {
            CIVector(x: extent.minX + value.x * extent.width, y: extent.minY + value.y * extent.height)
        }
        return image.applyingFilter("CIPerspectiveCorrection", parameters: [
            "inputTopLeft": point(rectangle.topLeft),
            "inputTopRight": point(rectangle.topRight),
            "inputBottomLeft": point(rectangle.bottomLeft),
            "inputBottomRight": point(rectangle.bottomRight)
        ])
    }

    private func compactEmbedding(_ image: CIImage) -> [Double] {
        let context = CIContext(options: [.workingColorSpace: NSNull()])
        let target = CGRect(x: 0, y: 0, width: 8, height: 8)
        let sourceWidth = max(CGFloat(1), image.extent.width)
        let sourceHeight = max(CGFloat(1), image.extent.height)
        let scale = CGAffineTransform(scaleX: CGFloat(8) / sourceWidth, y: CGFloat(8) / sourceHeight)
        let normalized = image.transformed(by: scale).cropped(to: target)
        var bytes = [UInt8](repeating: 0, count: 8 * 8 * 4)
        context.render(normalized, toBitmap: &bytes, rowBytes: 8 * 4, bounds: target, format: .RGBA8, colorSpace: CGColorSpaceCreateDeviceRGB())

        var embedding: [Double] = []
        embedding.reserveCapacity(64)
        for index in stride(from: 0, to: bytes.count, by: 4) {
            let red = Double(bytes[index])
            let green = Double(bytes[index + 1])
            let blue = Double(bytes[index + 2])
            let luminance = (0.299 * red + 0.587 * green + 0.114 * blue) / 255.0
            embedding.append(luminance)
        }
        return embedding
    }

    private func quality(_ image: CIImage, regions: [VNRectangleObservation]) -> [String: Any] {
        let coverage = regions.map { $0.boundingBox.width * $0.boundingBox.height }.max() ?? 0
        return ["blurScore": 0.5, "brightnessScore": 0.5, "glareScore": 0.0, "labelCoverage": coverage, "warnings": []]
    }

    private func elapsed(_ started: CFAbsoluteTime) -> Double { (CFAbsoluteTimeGetCurrent() - started) * 1000 }
    private func isCancelled(_ signalId: String?) -> Bool {
        guard let signalId else { return false }
        cancelLock.lock(); defer { cancelLock.unlock() }
        return cancelled.contains(signalId)
    }
    private func thermalStateName(_ state: ProcessInfo.ThermalState) -> String {
        switch state { case .nominal: return "nominal"; case .fair: return "fair"; case .serious: return "serious"; case .critical: return "critical"; @unknown default: return "unknown" }
    }
}
